hbase-server/src/main/java/org/apache/hadoop/hbase/master/normalizer/MergeNormalizer.java

```diff
@@ -0,0 +1,154 @@
+    List<NormalizationPlan> plans = new ArrayList<NormalizationPlan>();
+    List<HRegionInfo> tableRegions = masterServices.getAssignmentManager().getRegionStates().
+      getRegionsOfTable(table);
+    if (tableRegions == null || tableRegions.size() < MIN_REGION_COUNT) {
+      int nrRegions = tableRegions == null ? 0 : tableRegions.size();
+      LOG.debug("Table " + table + " has " + nrRegions + " regions, required min number"
+        + " of regions for normalizer to run is " + MIN_REGION_COUNT + ", not running normalizer");
+      return null;
+    }
+
+    LOG.debug("Computing normalization plan for table: " + table +
+      ", number of regions: " + tableRegions.size());
+
+    long totalSizeMb = 0;
+    int acutalRegionCnt = 0;
+
+    for (int i = 0; i < tableRegions.size(); i++) {
+      HRegionInfo hri = tableRegions.get(i);
+      long regionSize = getRegionSize(hri);
+      //don't consider regions that are in bytes for averaging the size.
+      if (regionSize > 0) {
+        acutalRegionCnt++;
+        totalSizeMb += regionSize;
+      }
+    }
+
+    double avgRegionSize = acutalRegionCnt == 0 ? 0 : totalSizeMb / (double) acutalRegionCnt;
+
+    LOG.debug("Table " + table + ", total aggregated regions size: " + totalSizeMb);
+    LOG.debug("Table " + table + ", average region size: " + avgRegionSize);
+
+    int candidateIdx = 0;
+    while (candidateIdx < tableRegions.size()) {
+      HRegionInfo hri = tableRegions.get(candidateIdx);
+      long regionSize = getRegionSize(hri);
+      if (candidateIdx == tableRegions.size() - 1) {
+        break;
+      }
+      HRegionInfo hri2 = tableRegions.get(candidateIdx + 1);
+      long regionSize2 = getRegionSize(hri2);
+      if (regionSize >= 0 && regionSize2 >= 0 && regionSize + regionSize2 < avgRegionSize) {
+        Timestamp hriTime = new Timestamp(hri.getRegionId());
+        Timestamp hri2Time = new Timestamp(hri2.getRegionId());
+        Timestamp currentTime = new Timestamp(System.currentTimeMillis());
+        try {
+          // atleast one of the two regions should be older than MIN_REGION_DURATION days
+          if (!(new Timestamp(hriTime.getTime() + TimeUnit.DAYS.toMillis(MIN_DURATION_FOR_MERGE)))
+              .after(currentTime)
+              || !(new Timestamp(
+                  hri2Time.getTime() + TimeUnit.DAYS.toMillis(MIN_DURATION_FOR_MERGE)))
+                      .after(currentTime)) {
+            LOG.info(
+              "Table " + table + ", small region size: " + regionSize + " plus its neighbor size: "
+                  + regionSize2 + ", less than the avg size " + avgRegionSize + ", merging them");
+            plans.add(new MergeNormalizationPlan(hri, hri2));
+            candidateIdx++;
+          }
+        } catch (Exception e) {
+          e.printStackTrace();
+        }
+      }
+      candidateIdx++;
+    }
+    if (plans.isEmpty()) {
+      LOG.debug("No normalization needed, regions look good for table: " + table);
+      return null;
```

Comment:
If we return empty list, would that also have same impact as returning null?

Reply:
Interface recommendation is to return null if no action has to be taken. Well as per the logic we can pass empty list and nothing will break. Its equivalent to not doing anything.